README
NAME
    Mail::DMARC - Perl implementation of DMARC

VERSION
    version 1.20141119

SYNOPSIS
    DMARC: Domain-based Message Authentication, Reporting and Conformance

      my $dmarc = Mail::DMARC::PurePerl->new(
        ... # see the documentation for the "new" method for required args
      );

      my $result = $dmarc->validate();

     if ( $result->result eq 'pass' ) {
         ...continue normal processing...
         return;
     };

     # any result that did not pass is a fail. Now for disposition

     if ( $result->evalated->disposition eq 'reject' ) {
         ...treat the sender to a 550 ...
     };
     if ( $result->evalated->disposition eq 'quarantine' ) {
         ...assign a bunch of spam points...
     };
     if ( $result->evalated->disposition eq 'none' ) {
         ...continue normal processing...
     };

DESCRIPTION
    This module is a suite of tools for implementing DMARC. It adheres very
    tightly to the 2013 DMARC draft, intending to implement every MUST and
    every SHOULD.

    This module can be used...

    *   by MTAs and filtering tools like SpamAssassin to validate that
        incoming messages are aligned with the purported sender's policy.

    *   by email senders, to receive DMARC reports from other mail servers
        and display them via CLI and web interfaces.

    *   by MTA operators to send DMARC reports to DMARC author domains.

    When a message arrives via SMTP, the MTA or filtering application can
    pass in a small amount of metadata about the connection (envelope
    details, SPF and DKIM results) to Mail::DMARC. When the validate method
    is called, Mail::DMARC will determine if:

     a. the header_from domain exists
     b. the header_from domain publishes a DMARC policy
     c. if not, end processing
     d. does the message conform to the published policy?
     e. did the policy request reporting? If so, save details.

    The validation results are returned as a Mail::DMARC::Result object. If
    the author domain requested a report, it was saved to the Report Store.
    The Store class includes a SQL implementation that is tested with SQLite
    and MySQL.

    There is more information available in the $result object. See
    Mail::DMARC::Result for complete details.

    Reports are viewed with the dmarc_view_reports program or with a web
    browser and the dmarc_httpd program.

    Aggregate reports are sent to their requestors with the
    dmarc_send_reports program.

    For aggregate reports that you have been sent, the dmarc_receive program
    will parse the email messages (from IMAP, Mbox, or files) and save the
    report results into the Report Store.

    The report store can use the same database to store reports you have
    received as well as reports you will send. There are several ways to
    identify the difference, including:

    *   received reports will have a null value for
        report_policy_published.rua

    *   outgoing reports will have null values for report.uuid and
        report_record.count

CLASSES
    Mail::DMARC - the perl interface for DMARC

    Mail::DMARC::Policy - a DMARC policy

    Mail::DMARC::PurePerl - Pure Perl implementation of DMARC

    Mail::DMARC::Result - the results of applying policy

    Mail::DMARC::Report - Reporting: the R in DMARC

      Mail::DMARC::Report::Send - send reports via SMTP & HTTP

      Mail::DMARC::Report::Receive - receive and store reports from email,
      HTTP

      Mail::DMARC::Report::Store - a persistent data store for aggregate
      reports

      Mail::DMARC::Report::View - CLI and CGI methods for viewing reports

    Mail::DMARC::libopendmarc
    <http://search.cpan.org/~shari/Mail-DMARC-opendmarc> - an XS
    implementation using libopendmarc

METHODS
  new
    Create a DMARC object.

        my $dmarc = Mail::DMARC::PurePerl->new;

    Populate it.

        $dmarc->source_ip('192.0.1.1');
        $dmarc->envelope_to('recipient.example.com');
        $dmarc->envelope_from('sender.example.com');
        $dmarc->header_from('sender.example.com');
        $dmarc->dkim( $dkim_verifier );
        $dmarc->spf(
            domain => 'example.com',
            scope  => 'mfrom',
            result => 'pass',
                );

    Run the request:

        my $result = $dmarc->validate();

    Alternatively, pass in all the required parameters in one shot:

        my $dmarc = Mail::DMARC::PurePerl->new(
                source_ip     => '192.0.1.1',
                envelope_to   => 'example.com',
                envelope_from => 'cars4you.info',
                header_from   => 'yahoo.com',
                dkim          => $dkim_results,  # same format
                spf           => $spf_results,   # as previous example
                );
        my $result = $dmarc->validate();

  source_ip
    The remote IP that attempted sending the message. DMARC only uses this
    data for reporting to domains that request DMARC reports.

  envelope_to
    The domain portion of the RFC5321.RcptTo, (aka, the envelope recipient),
    and the bold portion in the following example:

            RCPT TO:<user@example.com>

  envelope_from
    The domain portion of the RFC5321.MailFrom, (aka, the envelope sender).
    That is the the bold portion in the following example:

            MAIL FROM:<user@example.com>

  header_from
    The domain portion of the RFC5322.From, aka, the From message header.

            From: Ultimate Vacation <sweepstakes@example.com>

    You can instead pass in the entire From: header with header_from_raw.

  header_from_raw
    Retrieve the header_from domain by parsing it from a raw From
    field/header. The domain portion is extracted by get_dom_from_header,
    which is fast, generally effective, but also rather crude. It has
    limits, so read the description.

  dkim
    If Mail::DKIM::Verifier was used to validate the message, just pass in
    the Mail::DKIM::Verifier object that processed the message:

        $dmarc->dkim( $dkim_verifier );

    Otherwise, pass in an array reference. Each member of the DKIM array
    results represents a DKIM signature in the message and consists of the 4
    keys shown in this example:

        $dmarc->dkim( [
                {
                    domain      => 'example.com',
                    selector    => 'apr2013',
                    result      => 'fail',
                    human_result=> 'fail (body has been altered)',
                },
                {
                    # 2nd signature, if present
                },
            ] );

    The dkim results can also be build iteratively by passing in key value
    pairs or hash references for each signature in the message:

        $dmarc->dkim( domain => 'sig1.com', result => 'fail' );
        $dmarc->dkim( domain => 'sig2.com', result => 'pass' );
        $dmarc->dkim( { domain => 'example.com', result => 'neutral' } );

    Each hash or hashref is appended to the dkim array.

    The dkim result is an array reference.

   domain
    The d= parameter in the DKIM signature

   selector
    The s= parameter in the DKIM signature

   result
    The validation results of this signature. One of: none, pass, fail,
    policy, neutral, temperror, or permerror

   human result
    Additional information about the DKIM result. This is comparable to
    Mail::DKIM::Verifier->result_detail.

  spf
    The spf method works exactly the same as dkim. It accepts named
    arguments, a hashref, or an arrayref:

        $dmarc->spf(
            domain => 'example.com',
            scope  => 'mfrom',
            result => 'pass',
        );

    The SPF domain and result are required for DMARC validation and the
    scope is used for reporting.

   domain
    The SPF checked domain

   scope
    The scope of the checked domain: mfrom, helo

   result
    The SPF result code: none, neutral, pass, fail, softfail, temperror, or
    permerror.

DESIGN & GOALS
  Correct
    The DMARC spec is lengthy and evolving, making correctness a moving
    target. In cases where correctness is ambiguous, options are generally
    provided.

  Easy to use
    The effectiveness of DMARC will improve significantly as adoption
    increases. Proving an implementation of DMARC that SMTP utilities like
    SpamAssassin, amavis, and qpsmtpd can consume will aid adoption.

    The list of dependencies appears long because of reporting. If this
    module is used without reporting, the number of dependencies not
    included with perl is about 5. See the [Prereq] versus [Prereq /
    Recommends] sections in dist.ini.

  Maintainable
    Since DMARC is evolving, this implementation aims to be straight forward
    and dare I say, easy, to alter and extend. The programming style is
    primarily OO, which carries a small performance penalty but large
    dividends in maintainability.

    When multiple options are available, such as when sending reports via
    SMTP or HTTP, calls should be made to the parent Send class, to broker
    the request. When storing reports, calls are made to the Store class,
    which dispatches to the SQL class. The idea is that if someone desired a
    data store other than the many provided by perl's DBI class, they could
    easily implement their own. If you do, please fork it on GitHub and
    share.

  Fast
    If you deploy this in an environment where performance is insufficient,
    please profile the app and submit a report and preferably, patches.

SEE ALSO
    Mail::DMARC on GitHub: https://github.com/msimerson/mail-dmarc

    Mar 13, 2013 Draft:
    http://tools.ietf.org/html/draft-kucherawy-dmarc-base-00

    Mar 30, 2012 Draft: http://www.dmarc.org/draft-dmarc-base-00-02.txt

    Best Current Practices:
    http://tools.ietf.org/html/draft-crocker-dmarc-bcp-03

HISTORY
    The daddy of this perl module was a DMARC module for the qpsmtpd MTA.

    Qpsmtpd plugin:
    https://github.com/qpsmtpd-dev/qpsmtpd-dev/blob/master/plugins/dmarc

AUTHORS
    *   Matt Simerson <msimerson@cpan.org>

    *   Davide Migliavacca <shari@cpan.org>

CONTRIBUTORS
    *   Benny Pedersen <me@junc.eu>

    *   ColocateUSA.net <company@colocateusa.net>

    *   Ricardo Signes <rjbs@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by ColocateUSA.com.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes

1.20141119 2014-11-18 America/Los_Angeles

 - added auto_save option for validation reports
 - updated bin/install_deps.pl

1.20141030 2014-10-30 America/Los_Angeles

 - percent policy logic wasn't being applied correctly
 - fix for reasons not stored in SQL

1.20140711 2014-07-10 America/Los_Angeles

 - Store/SQL: use full sql name in WHERE clause
 - DMARC/HTTP: added error handling and tests
 - removed excess comma in mail_dmarc_schema.mysql
 - added quotes around ($commit || ''), just in case
 - try IMAP fetch without SORT if no results, for IMAP servers like Gmail that don't support SORT
 - warn but still pass test if DNS query fails

1.20140623 2014-06-23 America/Los_Angeles

 - updated tests to accomodate the cached PSL

1.20140622 2014-06-22 America/Los_Angeles

 - load PSL into hash to speed subsequent lookups (esp for daemon)
 - uncommented Net::Server in Prereqs/Recommended section
 - added INSTALL
 - updated dmarc_httpd description to note validation feature
 - updated public_suffix_list

1.20140210 2014-02-09 America/Los_Angeles

 - NEW FEATURE: added HTTP validation service (see dmarc_httpd)
 - install_deps: install optional prereqs by default
 - added Best Current Practices link on main page
 - minor tweaks to Pod (Ricardo Signes)
 - PurePerl: added comments about Sender header
    when message has multiple-address format used in the From header
 - updated public_suffix_list

1.20130906 2013-09-06 America/Los_Angeles

 - handle errors encountered when reporting address is illegal
 - delete reports that return a SMTP 5XX code for the recipient
 - delete reports after encountering 12 errors
 - added 'too big' notices when report size exceeds limit
 - updated install_deps.pl

1.20130625 2013-06-24 America/Los_Angeles

 - added a bunch of tests from http://dmarc-qa.com
 - URI: supress undef error if URI scheme not defined
 - policy->parse: properly parse records with unnecessary trailing ;
 - reporting is 'external' based on Org Domain (was email domain)

1.20130616 2013-06-16 America/Los_Angeles

 - combined update/replace SQL methods
 - dmarc_view_reports: fix duplicated variable name

1.20130615 2013-06-15 America/Los_Angeles

 - bug fixes and purge unused classes

1.20130614 2013-06-13 America/Los_Angeles

 - Added whitelist feature
 - SMTP: remove Subject: Report-ID <braces>
 - SMTP: more types of SMTP errors are stored and reported
 - dmarc_send_reports: added verbose option
 - dmarc_view_reports: fix for searches with MySQL backend

1.20130612 2013-06-11

 -  dmarc_view_reports: improve gentoo support by adding /usr
    to search path for GeoIP DBs on gentoo - Benny Pedersen

1.20130610 2013-06-10

 - tolerate receiving reports with no records (ahem, hotmail.com)
 - simplify SMTP flow-of-control, additional SMTP tests
 - avoid the join/split of binip in SQL::populate_agg_records
 - replace carp with warn in several places (more legible warning)
 - added RUA validity checks to dmarc_lookup

1.20130605 2013-06-05

 - in aggregate reports, group by IP and auth results (was only IP)
 - refactored SQL::retrieve_todo into 3 methods, added tests
 - SQL: added unique constraint on domain.domain

1.20130604 2013-06-04

 - main branches are master (devel) and releases (more obvious)
 - added mailing list impact FAQ
 - SQL: removed record.rcpt_dom
 - corrected a XML schema error
 - index.html
    - widened disposition column
    - only show rcpt domain in record (subgrid)
    - corrected subgrid row_id
 - additional validation of aggregate reports

1.20130601 2013-05-31

 - make sure a report record exists when fetching SMTP todo
 - added insecure SMTP fallback if STARTTLS fails
 - added color coded results to HTTP grid

1.20130531 2013-05-30

 - added gzip support to HTTP server, compressed JS files
 - reason is internally an arrayref of hashrefs (was a single hashref)
 - documentation additions
 - removed unused JS files
 - add validation and fixup of SPF result for incoming reports
 - normalized domain columns in spf & dkim tables

1.20130528 2013-05-28

 - bump major version to 1
 - normalized domain columns in report_record
 - fixups to handle reports with invalid formatting
 - improved handling for IMAP SSL connections
 - made internal represention of Mail::DMARC::dkim & spf consistent
   with their aggregate report representation

0.20130528 2013-05-27

 - updated Send/SMTP to use report::aggregate
 - switched back to gzip reports (instead of zip)
 - dmarc_view_reports, added filtering ability, GeoIP location

0.20130524 2013-05-23

 - added bin/dmarc_httpd
 - added bin/dmarc_view_reports
 - renamed: dmarc_report -> dmarc_send_reports

0.20130521 2013-05-21

 - check for report_record existence before insertion
 - SQL: added report_record.count column
 - subclassed aggregreate reports into Report::Aggregate
   - consolidates two agg. rep. generation methods to one
 - SQL: added table report_error
 - updated SQLite schema with native column types

0.20130520 2013-05-19

 - added bin/dmarc_receive (via IMAP, mbox, or message file)
 - added report retrieval via IMAP
 - extract sender domain from subject or MIME metadata
 - SQL: added author.extra_contact
 - SQL: removed 'NOT NULL' requirements for values often missing from incoming
        reports.

0.20130517 2013-05-17

 - send reports with zip until 7/1, gzip after
 - replace Socket 2 with Socket6 (better Windows compatibility)
 - added parsing of incoming email reports
 - added author and domain tables
 - added three related columns from/rcpt/author ids to report table
 - add email hostname to MX list when attempting SMTP delivery
 - during report delivery, check report URI max size

0.20130515 2013-05-15

 - use File::ShareDir to access share/*
 - added external reporting verification

0.20130514 2013-05-14

 - moved DNS settings into config file
 - fixed a case where disposition was not set
 - added bin/dmarc_report
   - sends email reports with Email::MIME & Net::SMTPS
 - deletes reports after successful delivery

lib/Mail/DMARC.pm
package Mail::DMARC;
our $VERSION = '1.20141119'; # VERSION
use strict;
use warnings;

use Carp;

use parent 'Mail::DMARC::Base';
require Mail::DMARC::Policy;
require Mail::DMARC::Report;
require Mail::DMARC::Result;

sub source_ip {
    return $_[0]->{source_ip} if 1 == scalar @_;
    croak "invalid source_ip" if !$_[0]->is_valid_ip( $_[1] );
    return $_[0]->{source_ip} = $_[1];
}

sub envelope_to {
    return $_[0]->{envelope_to} if 1 == scalar @_;
    croak "invalid envelope_to" if !$_[0]->is_valid_domain( $_[1] );
    return $_[0]->{envelope_to} = $_[1];
}

sub envelope_from {
    return $_[0]->{envelope_from} if 1 == scalar @_;
    croak "invalid envelope_from" if !$_[0]->is_valid_domain( $_[1] );
    return $_[0]->{envelope_from} = $_[1];
}

sub header_from {
    return $_[0]->{header_from} if 1 == scalar @_;
    croak "invalid header_from" if !$_[0]->is_valid_domain( $_[1] );
    return $_[0]->{header_from} = lc $_[1];
}

sub header_from_raw {
    return $_[0]->{header_from_raw} if 1 == scalar @_;
#croak "invalid header_from_raw: $_[1]" if 'from:' ne lc substr($_[1], 0, 5);
    return $_[0]->{header_from_raw} = lc $_[1];
}

sub local_policy {
    return $_[0]->{local_policy} if 1 == scalar @_;

    # TODO: document this, when and why it would be used
    return $_[0]->{local_policy} = $_[1];
}

sub dkim {
    my ( $self, @args ) = @_;
    return $self->{dkim} if 0 == scalar @args;

    if ( scalar @args > 1 ) {
        croak "invalid arguments to dkim" if @args % 2;
        push @{ $self->{dkim} }, { @args };
        $self->is_valid_dkim;
        return $self->{dkim};
    };

    my $dkim = shift @args;

    croak "invalid dkim argument" if ! ref $dkim;

    if ( ref $dkim eq 'Mail::DKIM::Verifier' ) {
        return $self->dkim_from_mail_dkim($dkim);
    };

    if ( 'ARRAY' eq ref $dkim ) {
        $self->{dkim} = $dkim;
        $self->is_valid_dkim;
        return $self->{dkim};
    }

    if ( 'HASH' eq ref $dkim ) {
        push @{ $self->{dkim} }, $dkim;
        $self->is_valid_dkim;
        return $self->{dkim};
    };

    croak "invalid dkim argument";
}

sub dkim_from_mail_dkim {
    my ( $self, $dkim ) = @_;

    # A DKIM verifier will have result and signature methods.
    foreach my $s ( $dkim->signatures ) {
        push @{ $self->{dkim} },
            {
            domain       => $s->domain,
            selector     => $s->selector,
            result       => $s->result,
            human_result => $s->result_detail,
            };
    }
    return $self->{dkim};
}

sub spf {
    my ( $self, @args ) = @_;
    return $self->{spf} if 0 == scalar @args;

    if ( scalar @args == 1 && ref $args[0] ) {
        if ( ref $args[0] eq 'HASH' ) {
            push @{ $self->{spf} }, $args[0];
            return $self->{spf};
        };
        if ( ref $args[0] eq 'ARRAY' ) {
            $self->{spf} = $args[0];
            return $self->{spf};
        }
    }

    croak "invalid arguments" if @args % 2;
    push @{ $self->{spf} }, {@args};
    $self->is_valid_spf();
    return $self->{spf};
}

sub policy {
    my ( $self, @args ) = @_;
    return $self->{policy} if ref $self->{policy} && 0 == scalar @args;
    return $self->{policy} = Mail::DMARC::Policy->new(@args);
}

sub report {
    my $self = shift;
    return $self->{report} if ref $self->{report};
    return $self->{report} = Mail::DMARC::Report->new();
}

sub result {
    my $self = shift;
    return $self->{result} if ref $self->{result};
    return $self->{result} = Mail::DMARC::Result->new();
}

sub is_subdomain {
    return $_[0]->{is_subdomain} if 1 == scalar @_;
    croak "invalid boolean" if 0 == grep {/^$_[1]$/ix} qw/ 0 1/;
    return $_[0]->{is_subdomain} = $_[1];
}

sub is_valid_dkim {
    my $self = shift;

    foreach my $dkim ( @{ $self->{dkim} } ) {
        foreach my $f (qw/ domain result /) {
            if ( !$dkim->{$f} ) {
                croak "DKIM value $f is required!";
            }
        }

        my @dkim_r = qw/ pass fail neutral none permerror policy temperror /;
        if ( !grep { $_ eq lc $dkim->{result} } @dkim_r ) {
            croak "invalid DKIM result!";
        }
    };
    return 1;
};

sub is_valid_spf {
    my $self = shift;

    foreach my $spf ( @{ $self->{spf} } ) {
        foreach my $f (qw/ domain result scope /) {
            if ( !$spf->{$f} ) {
                croak "SPF $f is required!";
            }
        }

        croak if $spf->{result} &&
            ! $self->is_valid_spf_result( $spf->{result} );

        croak if $spf->{scope} &&
            ! $self->is_valid_spf_scope( $spf->{scope} );

        if ( $spf->{result} =~ /^pass$/i && !$spf->{domain} ) {
            croak "SPF pass MUST include the RFC5321.MailFrom domain!";
        }
    };
    return 1;
}

sub save_aggregate {
    my ($self) = @_;

    my $agg = $self->report->aggregate;

    # put config information in report metadata
    foreach my $f ( qw/ org_name email extra_contact_info report_id / ) {
        $agg->metadata->$f( $self->config->{organization}{$f} );
    };
    $agg->metadata->begin( time );
    $agg->metadata->end( time + ($self->result->published->ri || 86400 ));

    $agg->policy_published( $self->result->published );
# could pass in $self as the identifier, and $self->result as the
# policy_evaluated. This documents what's being passed.
    $agg->record({
                row => {
                    source_ip         => $self->source_ip,
                    policy_evaluated  => {
                        disposition   => $self->result->disposition,
                        dkim          => $self->result->dkim,
                        spf           => $self->result->spf,
                        reason        => [ $self->result->reason ],
                    },
                },
                identifiers => {
                    envelope_to   => $self->envelope_to,
                    envelope_from => $self->envelope_from,
                    header_from   => $self->header_from,
                    },
                auth_results => {
                    dkim          => $self->dkim,
                    spf           => $self->spf,
                    },
            });

    return $self->report->save_aggregate;
};

1;

# ABSTRACT: Perl implementation of DMARC

=pod

=head1 NAME

Mail::DMARC - Perl implementation of DMARC

=head1 VERSION

version 1.20141119

=head1 SYNOPSIS

DMARC: Domain-based Message Authentication, Reporting and Conformance

  my $dmarc = Mail::DMARC::PurePerl->new(
    ... # see the documentation for the "new" method for required args
  );

  my $result = $dmarc->validate();

 if ( $result->result eq 'pass' ) {
     ...continue normal processing...
     return;
 };

 # any result that did not pass is a fail. Now for disposition

 if ( $result->evalated->disposition eq 'reject' ) {
     ...treat the sender to a 550 ...
 };
 if ( $result->evalated->disposition eq 'quarantine' ) {
     ...assign a bunch of spam points...
 };
 if ( $result->evalated->disposition eq 'none' ) {
     ...continue normal processing...
 };

=head1 DESCRIPTION

This module is a suite of tools for implementing DMARC. It adheres very tightly to the 2013 DMARC draft, intending to implement every MUST and every SHOULD.

This module can be used...

=over 4

=item *

by MTAs and filtering tools like SpamAssassin to validate that incoming messages are aligned with the purported sender's policy.

=item *

by email senders, to receive DMARC reports from other mail servers and display them via CLI and web interfaces.

=item *

by MTA operators to send DMARC reports to DMARC author domains.

=back

When a message arrives via SMTP, the MTA or filtering application can pass in a small amount of metadata about the connection (envelope details, SPF and DKIM results) to Mail::DMARC. When the B<validate> method is called, Mail::DMARC will determine if:

 a. the header_from domain exists
 b. the header_from domain publishes a DMARC policy
 c. if not, end processing
 d. does the message conform to the published policy?
 e. did the policy request reporting? If so, save details.

The validation results are returned as a L<Mail::DMARC::Result> object. If the author domain requested a report, it was saved to the L<Report Store|Mail::DMARC::Report::Store>. The Store class includes a SQL implementation that is tested with SQLite and MySQL.

There is more information available in the $result object. See L<Mail::DMARC::Result> for complete details.

Reports are viewed with the L<dmarc_view_reports> program or with a web browser and the L<dmarc_httpd> program.

Aggregate reports are sent to their requestors with the L<dmarc_send_reports> program.

For aggregate reports that you have been sent, the L<dmarc_receive> program will parse the email messages (from IMAP, Mbox, or files) and save the report results into the L<Report Store|Mail::DMARC::Report::Store>.

The report store can use the same database to store reports you have received as well as reports you will send. There are several ways to identify the difference, including:

=over 4

=item *

received reports will have a null value for report_policy_published.rua

=item *

outgoing reports will have null values for report.uuid and report_record.count

=back

=head1 CLASSES

L<Mail::DMARC> - the perl interface for DMARC

L<Mail::DMARC::Policy> - a DMARC policy

L<Mail::DMARC::PurePerl> - Pure Perl implementation of DMARC

L<Mail::DMARC::Result> - the results of applying policy

L<Mail::DMARC::Report> - Reporting: the R in DMARC

=over 2

L<Mail::DMARC::Report::Send> - send reports via SMTP & HTTP

L<Mail::DMARC::Report::Receive> - receive and store reports from email, HTTP

L<Mail::DMARC::Report::Store> - a persistent data store for aggregate reports

L<Mail::DMARC::Report::View> - CLI and CGI methods for viewing reports

=back

L<Mail::DMARC::libopendmarc|http://search.cpan.org/~shari/Mail-DMARC-opendmarc> - an XS implementation using libopendmarc

=head1 METHODS

=head2 new

Create a DMARC object.

    my $dmarc = Mail::DMARC::PurePerl->new;

Populate it.

    $dmarc->source_ip('192.0.1.1');
    $dmarc->envelope_to('recipient.example.com');
    $dmarc->envelope_from('sender.example.com');
    $dmarc->header_from('sender.example.com');
    $dmarc->dkim( $dkim_verifier );
    $dmarc->spf(
        domain => 'example.com',
        scope  => 'mfrom',
        result => 'pass',
            );

Run the request:

    my $result = $dmarc->validate();

Alternatively, pass in all the required parameters in one shot:

    my $dmarc = Mail::DMARC::PurePerl->new(
            source_ip     => '192.0.1.1',
            envelope_to   => 'example.com',
            envelope_from => 'cars4you.info',
            header_from   => 'yahoo.com',
            dkim          => $dkim_results,  # same format
            spf           => $spf_results,   # as previous example
            );
    my $result = $dmarc->validate();

=head2 source_ip

The remote IP that attempted sending the message. DMARC only uses this data for reporting to domains that request DMARC reports.

=head2 envelope_to

The domain portion of the RFC5321.RcptTo, (aka, the envelope recipient), and the bold portion in the following example:

=over 8

RCPT TO:<user@B<example.com>>

=back

=head2 envelope_from

The domain portion of the RFC5321.MailFrom, (aka, the envelope sender). That is the the bold portion in the following example:

=over 8

MAIL FROM:<user@B<example.com>>

=back

=head2 header_from

The domain portion of the RFC5322.From, aka, the From message header.

=over 8

From: Ultimate Vacation <sweepstakes@B<example.com>>

=back

You can instead pass in the entire From: header with header_from_raw.

=head2 header_from_raw

Retrieve the header_from domain by parsing it from a raw From field/header. The domain portion is extracted by L<get_dom_from_header|Mail::DMARC::PurePerl#get_dom_from_header>, which is fast, generally effective, but also rather crude. It has limits, so read the description.

=head2 dkim

If Mail::DKIM::Verifier was used to validate the message, just pass in the Mail::DKIM::Verifier object that processed the message:

    $dmarc->dkim( $dkim_verifier );

Otherwise, pass in an array reference. Each member of the DKIM array results represents a DKIM signature in the message and consists of the 4 keys shown in this example:

    $dmarc->dkim( [
            {
                domain      => 'example.com',
                selector    => 'apr2013',
                result      => 'fail',
                human_result=> 'fail (body has been altered)',
            },
            {
                # 2nd signature, if present
            },
        ] );

The dkim results can also be build iteratively by passing in key value pairs or hash references for each signature in the message:

    $dmarc->dkim( domain => 'sig1.com', result => 'fail' );
    $dmarc->dkim( domain => 'sig2.com', result => 'pass' );
    $dmarc->dkim( { domain => 'example.com', result => 'neutral' } );

Each hash or hashref is appended to the dkim array.

The dkim result is an array reference.

=head3 domain

The d= parameter in the DKIM signature

=head3 selector

The s= parameter in the DKIM signature

=head3 result

The validation results of this signature. One of: none, pass, fail, policy, neutral, temperror, or permerror

=head3 human result

Additional information about the DKIM result. This is comparable to Mail::DKIM::Verifier->result_detail.

=head2 spf

The spf method works exactly the same as dkim. It accepts named arguments, a hashref, or an arrayref:

    $dmarc->spf(
        domain => 'example.com',
        scope  => 'mfrom',
        result => 'pass',
    );

The SPF domain and result are required for DMARC validation and the scope is used for reporting.

=head3 domain

The SPF checked domain

=head3 scope

The scope of the checked domain: mfrom, helo

=head3 result

The SPF result code: none, neutral, pass, fail, softfail, temperror, or permerror.

=head1 DESIGN & GOALS

=head2 Correct

The DMARC spec is lengthy and evolving, making correctness a moving target. In cases where correctness is ambiguous, options are generally provided.

=head2 Easy to use

The effectiveness of DMARC will improve significantly as adoption increases. Proving an implementation of DMARC that SMTP utilities like SpamAssassin, amavis, and qpsmtpd can consume will aid adoption.

The list of dependencies appears long because of reporting. If this module is used without reporting, the number of dependencies not included with perl is about 5. See the [Prereq] versus [Prereq / Recommends] sections in dist.ini.

=head2 Maintainable

Since DMARC is evolving, this implementation aims to be straight forward and dare I say, easy, to alter and extend. The programming style is primarily OO, which carries a small performance penalty but large dividends in maintainability.

When multiple options are available, such as when sending reports via SMTP or HTTP, calls should be made to the parent Send class, to broker the request. When storing reports, calls are made to the Store class, which dispatches to the SQL class. The idea is that if someone desired a data store other than the many provided by perl's DBI class, they could easily implement their own. If you do, please fork it on GitHub and share.

=head2 Fast

If you deploy this in an environment where performance is insufficient, please profile the app and submit a report and preferably, patches.

=head1 SEE ALSO

Mail::DMARC on GitHub: https://github.com/msimerson/mail-dmarc

Mar 13, 2013 Draft: http://tools.ietf.org/html/draft-kucherawy-dmarc-base-00

Mar 30, 2012 Draft: http://www.dmarc.org/draft-dmarc-base-00-02.txt

Best Current Practices: http://tools.ietf.org/html/draft-crocker-dmarc-bcp-03

=head1 HISTORY

The daddy of this perl module was a DMARC module for the qpsmtpd MTA.

Qpsmtpd plugin: https://github.com/qpsmtpd-dev/qpsmtpd-dev/blob/master/plugins/dmarc

=head1 AUTHORS

=over 4

=item *

Matt Simerson <msimerson@cpan.org>

=item *

Davide Migliavacca <shari@cpan.org>

=back

=head1 CONTRIBUTORS

=over 4

=item *

Benny Pedersen <me@junc.eu>

=item *

ColocateUSA.net <company@colocateusa.net>

=item *

Ricardo Signes <rjbs@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by ColocateUSA.com.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

__END__
sub {}  # for vim automatic code folding

